packages/@lwc/module-resolver/src/index.ts

```diff
@@ -4,117 +4,147 @@
+    getLwcConfig,
     getModuleEntry,
     normalizeConfig,
-    loadConfig,
-    isString,
+    validateNpmConfig,
     mergeModules,
-    LWC_CONFIG_FILE,
 } from './utils';
 
-export interface RegistryEntry {
-    entry: string;
-    specifier: string;
-}
+function resolveModuleFromAlias(
+    moduleRecord: AliasModuleRecord,
+    opts: InnerResolverOptions
+): RegistryEntry {
+    const { name: specifier, path: modulePath } = moduleRecord;
+    const entry = path.resolve(opts.rootDir, modulePath);
+    if (!fs.existsSync(entry)) {
+        throw new Error(
+            `Unable to find AliasModuleRecord for "${specifier}". File ${entry} does not exist`
+        );
+    }
 
-export interface ModuleRegistryMap {
-    [key: string]: RegistryEntry;
-}
-export interface ModuleRecordObject {
-    name: string;
-    path: string;
+    return createRegistryEntry(entry, specifier, opts);
 }
 
-export interface ModuleResolverConfig {
-    rootDir: string;
-    modules: ModuleRecord[];
-}
+function resolveModuleFromDir(
+    specifier: string,
+    moduleRecord: DirModuleRecord,
+    opts: InnerResolverOptions
+): RegistryEntry | undefined {
+    const { dir } = moduleRecord;
+    const absModuleDir = path.isAbsolute(dir) ? dir : path.join(opts.rootDir, dir);
+    const [ns, name] = specifier.split('/');
 
-export type ModuleRecord = string | ModuleRecordObject;
-export interface LwcConfig {
-    modules: ModuleRecord[];
+    if (ns && name) {
+        const moduleDir = path.join(absModuleDir, ns, name);
+        const entry = getModuleEntry(moduleDir, name);
+        if (entry) {
+            return createRegistryEntry(entry, specifier, opts);
+        }
+    }
 }
 
-function resolveModulesFromDir(modulesDir: string): RegistryEntry[] {
-    const namespaces = fs.readdirSync(modulesDir);
-    const resolvedModules: RegistryEntry[] = [];
-    namespaces.forEach(ns => {
-        if (ns[0] !== '.' && fs.lstatSync(path.join(modulesDir, ns)).isDirectory()) {
-            const namespacedModuleDir = path.join(modulesDir, ns);
-            const modules = fs.readdirSync(namespacedModuleDir);
-            modules.forEach(moduleName => {
-                const moduleDir = path.join(namespacedModuleDir, moduleName);
-                const entry = getModuleEntry(moduleDir, moduleName);
-                if (entry) {
-                    const specifier = `${ns}/${moduleName}`;
-                    resolvedModules.push({ entry, specifier });
+function resolveModuleFromNpm(
+    specifier: string,
+    npmModuleRecord: NpmModuleRecord,
+    opts: InnerResolverOptions
+): RegistryEntry | undefined {
+    const { npm, map: aliasMapping } = npmModuleRecord;
+    const pkgJsonPath = require.resolve(`${npm}/package.json`, { paths: [opts.rootDir] });
+    const packageDir = path.dirname(pkgJsonPath);
+    const lwcConfig = getLwcConfig(packageDir);
+    const modules = lwcConfig && lwcConfig.modules;
+
+    if (modules) {
+        validateNpmConfig(lwcConfig);
```

Comment:
I would recommend inlining the `expose` check here to avoid having to do `lwc.expose || []` just to please typescript. Because we know at this point that `expose` has to be defined.